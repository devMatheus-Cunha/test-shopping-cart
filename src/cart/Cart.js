export default class Cart {
  items = []
  getTotal() {
    return this.items.reduce((acc, item) => {
      return acc + item.quantity * item.product.price
    }, 0)
  }

  addItem(item) {
    this.items.push(item)
  }

}
