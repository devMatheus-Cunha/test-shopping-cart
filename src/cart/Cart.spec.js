import Cart from './Cart';

describe('Cart', () => {
  let cart;
  let productOne = {
    title: 'Adidas',
    price: 120,
  };
  let productTwo = {
    title: 'Nike',
    price: 100,
  };

  // before each test
  beforeEach(() => {
    cart = new Cart();
  });

  describe('getTotal(', () => {
    it('should return 0 when getTotal() is executed in a newly created instace ', () => {
      expect(cart.getTotal().getAmount()).toEqual(0);
    });

    it('should multiply quantity and price and receive the total amount', () => {
      const item = {
        product: productOne,
        quantity: 2,
      };
      cart.add(item);

      expect(cart.getTotal().getAmount()).toEqual(240);
    });

    it('should ensure no more than on product quantity exists at a time', () => {
      cart.add({
        product: productOne,
        quantity: 2,
      });
      cart.add({
        product: productOne,
        quantity: 1,
      });

      expect(cart.getTotal().getAmount()).toEqual(240);
    });

    it('should update total when a product gets included then removed ', () => {
      cart.add({
        product: productOne,
        quantity: 2,
      });
      cart.add({
        product: productTwo,
        quantity: 1,
      });

      cart.remove(productOne);

      expect(cart.getTotal().getAmount()).toEqual(100);
    });
  });

  describe('checkout', () => {
    it('should return object with the total and the list of items', () => {
      cart.add({
        product: productOne,
        quantity: 2,
      });
      cart.add({
        product: productTwo,
        quantity: 5,
      });

      expect(cart.checkout()).toMatchSnapshot();
    });

    it('should return object with the total and the list of items when summary is called', () => {
      cart.add({
        product: productOne,
        quantity: 2,
      });
      cart.add({
        product: productTwo,
        quantity: 5,
      });

      expect(cart.summary()).toMatchSnapshot();
      expect(cart.getTotal().getAmount()).toBeGreaterThan(0)
    });

    it('should reset the cart  when checkout() is called', () => {
      cart.add({
        product: productTwo,
        quantity: 5,
      });
      cart.checkout()

      expect(cart.getTotal().getAmount()).toEqual(0);
    });
  });

  describe('special conditios', () => {
    it('should apply percentage discount quantity above minimum is passed ', () => {
      const condition = {
        percentage: 30,
        minimum: 2
      }

      cart.add({
        product: productOne,
        condition,
        quantity: 3
      })

      expect(cart.getTotal().getAmount()).toEqual(252)
    });
  });
});
