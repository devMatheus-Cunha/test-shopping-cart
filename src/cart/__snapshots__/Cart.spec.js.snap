// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cart checkout should return object with the total and the list of items 1`] = `
Object {
  "items": Array [
    Object {
      "product": Object {
        "price": 120,
        "title": "Adidas",
      },
      "quantity": 2,
    },
    Object {
      "product": Object {
        "price": 100,
        "title": "Nike",
      },
      "quantity": 5,
    },
  ],
  "total": 740,
}
`;

exports[`Cart checkout should return object with the total and the list of items when summary is called 1`] = `
Object {
  "items": Array [
    Object {
      "product": Object {
        "price": 120,
        "title": "Adidas",
      },
      "quantity": 2,
    },
    Object {
      "product": Object {
        "price": 100,
        "title": "Nike",
      },
      "quantity": 5,
    },
  ],
  "total": 740,
}
`;
